core/src/test/java/org/elasticsearch/common/cache/CacheTests.java

```diff
@@ -502,6 +502,51 @@ public void testComputeIfAbsentThrowsExceptionIfLoaderReturnsANullValue() {
         }
     }
 
+    public void testCachePollution() throws InterruptedException {
+        int numberOfThreads = randomIntBetween(2, 200);
+        final Cache<Integer, String> cache = CacheBuilder.<Integer, String>builder().build();
+        CountDownLatch latch = new CountDownLatch(1 + numberOfThreads);
+        List<Thread> threads = new ArrayList<>();
+        for (int i = 0; i < numberOfThreads; i++) {
+            Thread thread = new Thread(() -> {
+                latch.countDown();
+                Random random = new Random(random().nextLong());
+                for (int j = 0; j < numberOfEntries; j++) {
+                    Integer key = random.nextInt(numberOfEntries);
+                    boolean first;
+                    boolean second;
+                    do {
+                        first = random.nextBoolean();
+                        second = random.nextBoolean();
+                    } while (first && second);
+                    if (first && !second) {
+                        try {
+                            cache.computeIfAbsent(key, k -> {
+                                if (random.nextBoolean()) {
+                                    return Integer.toString(k);
+                                } else {
+                                    throw new Exception();
+                                }
+                            });
+                        } catch (ExecutionException e) {
+
+                        }
```

Comment:
Race condition here on our comments too. :)

Yes, your [second comment](https://github.com/elastic/elasticsearch/pull/14315/files#r43272120) is exactly the issue and I agree that we should add an assertion that we only ended up there because of an expected exception on the loader. I've added such an assertion in c8818e4f5ebfb5eea3e73a75c60f8b3ea8f77491.